### PR TITLE
Use shared storage in protocol

### PR DESCRIPTION
**Pull request by gterzian**
Part of https://github.com/massalabs/massa/issues/2178

### CODE REVIEW

File: massa-protocol-worker/src/protocol_worker.rs [hunk 1]
```diff
@@ -1129,23 +1136,37 @@ impl ProtocolWorker {
...
+                stored_block.block.header.clone(),
+                stored_block.block.operations.clone(),
+                stored_block
+                    .block
+                    .header
+                    .content
+                    .operation_merkle_root
+                    .clone(),
+                stored_block.block.header.content.slot,
+            )
```
Comment (gterzian) [inline massa-protocol-worker/src/protocol_worker.rs:1152]: Note the clones here are I believe mostly a result of being unable to hold a lock across async points, see https://github.com/massalabs/massa/issues/2179, so removing them could be a good follow-up.

File: massa-consensus-worker/src/consensus_worker.rs [hunk 2]
```diff
@@ -1103,7 +1103,27 @@ impl ConsensusWorker {
                     "consensus.consensus_worker.process_protocol_event.get_blocks",
                     { "list": list }
                 );
-                // TODO: use shared storage.
```
Comment (AurelienFT) [inline massa-consensus-worker/src/consensus_worker.rs:1106]: I tried to follow every branch of this code including message send and receive through the network and didn't see a store or retrieve call. Am I missing something ?

File: massa-protocol-worker/src/protocol_worker.rs [hunk 3]
```diff
@@ -1377,15 +1396,16 @@ impl ProtocolWorker {
...
                 if let Some((block_id, operation_set, endorsement_ids)) =
                     self.note_block_from_node(&block, &from_node_id).await?
                 {
                     let mut set = Set::<BlockId>::with_capacity_and_hasher(1, BuildMap::default());
                     set.insert(block_id);
                     self.stop_asking_blocks(set)?;
                     self.send_protocol_event(ProtocolEvent::ReceivedBlock {
-                        block_id,
-                        block,
+                        block: block_id,
```
Comment (AurelienFT) [inline massa-protocol-worker/src/protocol_worker.rs:1408]: maybe change the parameters to block_id also.

File: massa-consensus-worker/src/consensus_worker.rs (hunk 2 above)
Comment (gterzian) [reply massa-consensus-worker/src/consensus_worker.rs:1106]: I think the store is done in network in this PR: https://github.com/massalabs/massa/pull/2305

retrieve is done in various places, for example in `block_graph` in this PR https://github.com/massalabs/massa/pull/2233

**Comment by gterzian** [comment]
@AurelienFT I think here is makes sense to change `block` to `block_id` as you suggest, do you want to do this all in one commit and push to this branch?

**Comment by AurelienFT** [comment]
> @AurelienFT I think here is makes sense to change `block` to `block_id` as you suggest, do you want to do this all in one commit and push to this branch?

On the file tab you can select "Add suggestion on the batch" for all suggestions so that they are made in 1 commit

**Review by AurelienFT — APPROVED** [review]

File: massa-protocol-exports/src/protocol_controller.rs [hunk 4]
```diff
@@ -46,8 +45,7 @@ pub enum ProtocolPoolEvent {
     },
 }
 
-type BlocksResults =
-    Map<BlockId, Option<(Block, Option<Set<OperationId>>, Option<Vec<EndorsementId>>)>>;
+type BlocksResults = Map<BlockId, Option<(Option<Set<OperationId>>, Option<Vec<EndorsementId>>)>>;
```
[the remark this answers is not in the document]
Comment (gterzian) [reply massa-protocol-exports/src/protocol_controller.rs:48]: https://github.com/massalabs/massa/pull/2304/files#diff-8615507c1ee829f75d67206a9a0dd2238e744316412fa73c233db6cb91c4ced4R122

**Comment by gterzian** [comment]
@AurelienFT Thank you for your commit, actually the change had to be made in a few other places, so I force-pushed a commit that does it in those other places as well.

File: massa-graph/src/block_graph.rs [hunk 5]
```diff
@@ -1404,14 +1403,21 @@ impl BlockGraph {
         if self.genesis_hashes.contains(&block_id) {
             return Ok(());
         }
-        let slot = block.header.content.slot;
-        debug!("received block {} for slot {}", block_id, slot);
-        massa_trace!("consensus.block_graph.incoming_block", {"block_id": block_id, "block": block});
+
+        let slot = {
+            let stored_block = self.storage.retrieve_block(&block_id).unwrap();
+            let stored_block = stored_block.read();
```
[the remark this answers is not in the document]
Comment (AureliaDolo) [reply massa-graph/src/block_graph.rs:1409]: Thanks

**Pull request merged**